presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java

```diff
@@ -455,9 +456,14 @@ public synchronized QueryResults getNextResult(OptionalLong token, UriInfo uriIn
             data = ImmutableSet.of(ImmutableList.of(true));
         }
 
-        // only return a next if the query is not done or there is more data to send (due to buffering)
+        // only return a next if
+        // (1) the query is not done AND (query is not a already failed query or user configure to wait for stats for already-failed query)
+        //   OR
+        // (2)there is more data to send (due to buffering)
         URI nextResultsUri = null;
-        if (!queryInfo.isFinalQueryInfo() || !exchangeClient.isClosed()) {
+        if (!queryInfo.isFinalQueryInfo()
+                && (!queryInfo.getState().equals(QueryState.FAILED) || isWaitForQueryStatsAfterFailed(session))
+                || !exchangeClient.isClosed()) {
             nextResultsUri = createNextResultsUri(scheme, uriInfo);
```

Comment:
This is the core change I made;
The wait logic only change for `FAILED` state while remain the same for other states;
In below two circumstance, the `nextResutsUri`  will be present for user to further update query stats:
---- 
the query is not done AND (query is not a already failed query or user configure to wait for stats for already-failed query)
OR
there is more data to send (due to buffering)
----
It means that, when `queryInfo.isFinalQueryInfo()` return false, as long as the queryState is FAILED AND user configured to not-wait for final status , we set `nextResultsUri` to null , query finished with error;
@dain what's your advice?